[DPE-3340] README (#12)
canonical/central-uploader: README.md

# central-uploader

The `central-uploader` tool is to enable application release artifacts to Launchpad project pages.

# How it works

In order to be able to use the `central-uploader`, we need a Launchpad project, with a build proces,
producing artifacts as described in the **Project Built from Source (prerequisite)** section below.

The `centra-uploader`'s task in the chain is to download the existing artifacts from the
strictly protected Launchpad SOSS space, and publish them to the project's public Launchpad releases space.

## Registering a project

> If you want to register your project with the `central-uploader`, you need to add a section

Reviewer: I would just add this at the beginning:

> In order to enable upload of built artifacts to any of the platforms, the relevant project (repository and branch) needs to be registered in the `central-uploader`, such that new CI built are monitored for. Registering of a project can be done by adding a new element to the `products.json` (for software distributions) or `library.json` (for simple JARs).

> as below to the `products.json` file, and run the pipelines.

Reviewer: Maybe it is worth saying that rather than "you need to add a section" I would say that people should "raise a PR" :smile:

```
    {
        "name": "opensearch",
        "lp-building-project": "soss",
        "lp-releasing-project": "opensearch-releases",
        "lp-building-repo": "~data-platform/soss/+source/opensearch-build/+git/opensearch-build/",
        "lp-building-branch-prefix": "lp-2",
        "lp-consumer-key": "LP_CONSUMER_KEY",
        "lp-access-token": "LP_ACCESS_TOKEN",
        "lp-access-secret": "LP_SECRET_TOKEN",
        "artifactory-url": "https://canonical.jfrog.io/artifactory/dataplatform-opensearch-maven-local/",
        "artifactory-user": "ARTIFACTORY_USER",
        "artifactory-token": "ARTIFACTORY_TOKEN",
        "tarball-regex": "opensearch-*.tar.gz"
    }
```

Note that

>  - the `lp-building-branch-prefix` has to correspond to the release branch naming terminology of

Reviewer: ```suggestion
 - the `lp-building-branch-prefix` has to correspond to the release branch naming terminology of the repository
```

>    of the Project's SOSS build repository.

Reviewer: ```suggestion
 
```

 - `tarball-regex` has to correspond to the artifactory output name
 - the `lp-consumer-key`, `lp-access-token`, `lp-access-secret`, `artifactory-user`, `artifactory-token` fields
   are to be defined as in the example above

## Publishing a release

There are regular executions of the `central-uploader` Release pipelines, but they can also be triggered manually.

Within this pipeline 

>  1. Existing artifacts are downloaded from the project's Launchpad SOSS build space

Reviewer: ```suggestion
 1. Existing artifacts are downloaded from the Launchpad built
```

 2. A new Github Release is created for the project within the scope of the `central-uploader` repository
 3. A new tag is also created with the release here

>  4. The artifacts are uploaded to the project's public Launchpad releases space

Reviewer: 5. (Optional) The jars are uploaded to the Artifactory

## Testing

When triggering your first pipeline runs, keep in mind that a release is only to be published once.
If you would like to re-run the publishing method for testing purposes, you need to make sure that

> starting from item 2. of the list above all traces of the previous publish attemt under the same version are removed.

Reviewer: I'd say that if someone wants to test this, it is best to fork the repository and disable publishing to LaunchPad, just testing the publishing to GitHub to start with. We should not trigger the pipeline in the central uploader for testing.

# Project Built from Source (prerequisite)

In order to be able to use `central-uploader`, you need to make sure that your project complies to the following criteria.

### 0. Launchpad public git repo (optional)

Your project's source code has to reside in a Launchpad git repository

In case you may have multiple repositories to be bundled up in a single build, those repositories have to reside in

> the Launchpad public git space, available to anyone who has a Launchpad account

Reviewer: This is quite specific to opensearch. Openseach during build needs to clone the other repositories. Since LP does not support secrets, those repositories need to be public. 

But honestly, this step is really not that important for the github central uploader. It is more of an OpenSearch LaunchPad build process specifics

In case your project repository is the same space where you want to apply your build mechanism, you can skip this step.

Example: https://code.launchpad.net/~medib/opensearch-project-components/+git/opensearch-performance-analyzer

### 1. Launchpad build repository

> The repository where the build process is to be performed has to reside in the Launchpad SOSS space.

Reviewer: This is also quite use-case specific. I mean: we need to put the repository in this space because we want to use a canonical Jfrog, and the credentials for accessing to this are only injected by the LP team into the projects of the SOSS space. 

But again, this is not really relevant to the central uploader.

This area is strictly managed within Canonical Launchpad Team, and is created on request.
Access is similalry restricted.

Example (that only Canonical Data Platform Team can access): 

You need to define a well-distinguishable branch for your build project (for example using the version number of the project).

### 2. Launchpad build pipeline

> Within your SOSS build repo, a `.launchpad.yaml` file (similar to GitHub pipeline declaratinos) has to exist, with steps

Reviewer: Also this is not really relevant for the central uploader. It is how builds work in LP.
